truncate/table_schema.go

```diff
@@ -137,38 +123,46 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
 	return tables, nil
 }
 
-func fetchIndexSchemas(ctx context.Context, client *spanner.Client) ([]*indexSchema, error) {
-	// This query fetches defined indexes.
-	iter := client.Single().Query(ctx, spanner.NewStatement(`
-		SELECT INDEX_NAME, TABLE_NAME, PARENT_TABLE_NAME FROM INFORMATION_SCHEMA.INDEXES
-		WHERE INDEX_TYPE = 'INDEX' AND TABLE_CATALOG = '' AND TABLE_SCHEMA = '';
-	`))
+// filterTableSchemas filters tables by excludeTables.
+// If an exclude table is cascade deletable, its parent table is also excluded.
+func filterTableSchemas(tables []*tableSchema, excludeTables []string) []*tableSchema {
+	if len(excludeTables) == 0 {
+		return tables
+	}
 
-	var indexes []*indexSchema
-	if err := iter.Do(func(r *spanner.Row) error {
-		var (
-			indexName     string
-			baseTableName string
-			parent        spanner.NullString
-		)
-		if err := r.Columns(&indexName, &baseTableName, &parent); err != nil {
-			return err
+	excludes := make(map[string]struct{}, len(tables))
+	for _, t := range excludeTables {
+		excludes[t] = struct{}{}
+	}
+
+	// Add parent tables that may delete the exclude tables in cascade
+	// Since interleave tables can be hierarchical, tracing up to the top level is needed.
+	for {
+		excludeParents := make(map[string]struct{}, len(excludes))
+		for _, t := range tables {
+			if _, ok := excludes[t.tableName]; ok && t.isCascadeDeletable() {
+				if _, alreadyExcluded := excludes[t.parentTableName]; !alreadyExcluded {
+					excludeParents[t.parentTableName] = struct{}{}
+				}
+			}
 		}
 
-		var parentTableName string
-		if parent.Valid {
-			parentTableName = parent.StringVal
+		// loop until no more parent tables to exclude
+		if len(excludeParents) == 0 {
+			break
 		}
 
-		indexes = append(indexes, &indexSchema{
-			indexName:       indexName,
-			baseTableName:   baseTableName,
-			parentTableName: parentTableName,
-		})
-		return nil
-	}); err != nil {
-		return nil, err
+		for p := range excludeParents {
+			excludes[p] = struct{}{}
+		}
+	}
```

Comment:
This part might be a little less readable. 
Previous commits did not support hierarchical interleaving. Therefore, I have added a process that traces back to the top level hierarchy.

Reply:
Yeah, let me review this once another part is changed.